java-checks/src/main/java/org/sonar/java/checks/synchronization/ValueBasedObjectUsedForLockCheck.java

```diff
@@ -0,0 +1,85 @@
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.checks.synchronization;
+
+import com.google.common.collect.ImmutableSet;
+
+import org.sonar.check.Rule;
+import org.sonar.java.JavaVersionAwareVisitor;
+import org.sonar.plugins.java.api.IssuableSubscriptionVisitor;
+import org.sonar.plugins.java.api.JavaVersion;
+import org.sonar.plugins.java.api.semantic.Type;
+import org.sonar.plugins.java.api.tree.ExpressionTree;
+import org.sonar.plugins.java.api.tree.SynchronizedStatementTree;
+import org.sonar.plugins.java.api.tree.Tree;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Set;
+
+@Rule(key = "S3436")
+public class ValueBasedObjectUsedForLockCheck extends IssuableSubscriptionVisitor implements JavaVersionAwareVisitor {
+
+  private static final Set<String> VALUE_BASED_TYPES = ImmutableSet.of(
+    "java.time.chrono.HijrahDate",
+    "java.time.chrono.JapaneseDate",
+    "java.time.chrono.MinguoDate",
+    "java.time.chrono.ThaiBuddhistDate",
+    "java.util.Optional",
+    "java.util.OptionalDouble",
+    "java.util.OptionalLong",
+    "java.util.OptionalInt");
+
+  private static final String JAVA_TIME_PACKAGE = "java.time.";
+  private static final String JAVA_TIME_CLOCK = "java.time.Clock";
+
+
+  @Override
+  public boolean isCompatibleWithJavaVersion(JavaVersion version) {
+    return version.isJava8Compatible();
+  }
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return Collections.singletonList(Tree.Kind.SYNCHRONIZED_STATEMENT);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    ExpressionTree expression = ((SynchronizedStatementTree) tree).expression();
+    if (isValueBasedType(expression.symbolType())) {
+      reportIssue(expression, "Synchronize on some other object." + context.getJavaVersion().java8CompatibilityMessage());
+    }
+  }
+
+  private static boolean isValueBasedType(Type type) {
+    if (type.isUnknown() || type.is(JAVA_TIME_CLOCK)) {
+      return false;
+    }
+    if (VALUE_BASED_TYPES.stream().anyMatch(type::is)) {
+      return true;
+    }
+    String fullyQualifiedName = type.fullyQualifiedName();
+    return fullyQualifiedName.startsWith(JAVA_TIME_PACKAGE)
```

Comment:
I would rather go with a regexp `java\.time\.[^\.]` rather than startWith AND a regexp. This would allow to have a simpler code as well : `return anyMatch... && pattern.matcher(type.fqn()).matches();`